tests/test_shell_launcher.py

```diff
@@ -0,0 +1,323 @@
+    return tmp_dir
+
+
+def generate_output_files(tmp_dir: pathlib.Path):
+    """Generates output and error files within the run directory for testing."""
+    out_file = tmp_dir / "tmp.out"
+    err_file = tmp_dir / "tmp.err"
+    return out_file, err_file
+
+
+def generate_directory(test_dir: str):
+    """Generates a execution directory, output file, and error file for testing."""
+    execution_dir = create_directory(os.path.join(test_dir, "tmp"))
+    out_file, err_file = generate_output_files(execution_dir)
+    return execution_dir, out_file, err_file
+
+
+@pytest.fixture
+def shell_launcher():
+    launcher = ShellLauncher()
+    yield launcher
+    if any(proc.poll() is None for proc in launcher._launched.values()):
+        raise ("Test leaked processes")
+
+
+@pytest.fixture
+def shell_cmd(test_dir: str) -> ShellLauncherCommand:
+    """Fixture to create an instance of Generator."""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    return ShellLauncherCommand(
+        {}, run_dir, out_file, err_file, EchoHelloWorldEntity().as_program_arguments()
+    )
+
+
+# UNIT TESTS
+
+
+def test_shell_launcher_command_init(shell_cmd: ShellLauncherCommand, test_dir: str):
+    """Test that ShellLauncherCommand initializes correctly"""
+    assert shell_cmd.env == {}
+    assert shell_cmd.path == pathlib.Path(test_dir) / "tmp"
+    assert shell_cmd.stdout == shell_cmd.path / "tmp.out"
+    assert shell_cmd.stderr == shell_cmd.path / "tmp.err"
+    assert shell_cmd.command_tuple == EchoHelloWorldEntity().as_program_arguments()
+
+
+def test_shell_launcher_init(shell_launcher):
+    """Test that ShellLauncher initializes correctly"""
+    assert shell_launcher._launched == {}
+
+
+def test_shell_launcher_start_calls_popen(
+    shell_launcher, shell_cmd: ShellLauncherCommand
+):
+    """Test that the process leading up to the shell launcher popen call was correct"""
+    with unittest.mock.patch(
+        "smartsim._core.shell.shellLauncher.sp.Popen"
+    ) as mock_open:
+        _ = shell_launcher.start(shell_cmd)
+        mock_open.assert_called_once()
+
+
+def test_shell_launcher_start_calls_popen_with_value(
+    shell_launcher, shell_cmd: ShellLauncherCommand
+):
+    """Test that popen was called with correct values"""
+    with unittest.mock.patch(
+        "smartsim._core.shell.shellLauncher.sp.Popen"
+    ) as mock_open:
+        _ = shell_launcher.start(shell_cmd)
+        mock_open.assert_called_once_with(
+            shell_cmd.command_tuple,
+            cwd=shell_cmd.path,
+            env=shell_cmd.env,
+            stdout=shell_cmd.stdout,
+            stderr=shell_cmd.stderr,
+        )
+
+
+def test_popen_returns_popen_object(shell_launcher, test_dir: str):
+    """Test that the popen call returns a popen object"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
```

Comment:
Files look like they are opened, but unused

Reply:
its my favorite thing to do i think